### DISCUSSION TITLE
Does the async profiler signal the process reaper thread?

### DISCUSSION THREAD

**Discussion by mlichtblau** [async-profiler/async-profiler#782] [Q&A] [answered]
Hey, I have a follow up question about [this old comment](https://github.com/async-profiler/async-profiler/issues/279#issuecomment-784471241).

> If a busy thread is signaled while being in kernel, it will finish current system call and will proceed with a signal handler afterwards.

I wonder if this is also true for the [waitpid](https://linux.die.net/man/2/waitpid) call which from what I understand should return immediately with `errno` being set to `EINT` ([Stackoverflow Link](https://stackoverflow.com/questions/55412365/what-does-wait-and-waitpid-are-always-interrupted-when-a-signal-is-caught-mean) citing Advanced Programming in the Unix Environment)?

This is relevant as the JDK [spawns a process reaper thread](https://github.com/openjdk/jdk/blob/74121930e33686d2452170554776c0901f622d3e/src/java.base/share/classes/java/lang/ProcessHandleImpl.java#L155) when a process is started using the ProcessBuilder. This thread is [calling into native code](https://github.com/openjdk/jdk/blob/74121930e33686d2452170554776c0901f622d3e/src/java.base/unix/native/libjava/ProcessHandleImpl_unix.c#L238) and executing `waitpid`. 

My theory is that this thread should normally not consume any CPU time, but as the async-profiler is sending SIGPROF signals to that thread the loop is interrupted and errno set to `EINTR`. This is handled so it doesn't affect behavior but costs CPU time if you have many thousand child processes spawned.

Do you think this theory is possible and if yes, do you see anything that could be done to avoid this problem?
Thanks in advance!

**Comment by apangin** [comment, marked as answer]
async-profiler does not send SIGPROF signals.

- In `cpu` mode, `SIGPROF` is sent by the Operating System to a thread that has spent the specified amount of CPU time. If a thread is idle (like sleeping in `waitpid`), it obviously does not consume CPU and therefore will not be interrupted by `SIGPROF`.
- In `wall` mode, async-profiler samples all threads, both running and idle, by sending `SIGVTALRM` to them every given time interval. The target thread wakes up, records sample and retries a syscall if needed (no matter if it is `waitpid`, `poll`, `futex` or any other blocking syscall). This takes, let's say, 100 microseconds. If we sample with 100 millisecond interval, the overhead will be around 0.1%. Is it a problem? I don't think so. In any case, the overhead is directly manageable by adjusting the sampling interval: e.g., if you want to decrease the profiler overhead by 10x, increase the interval by the same ratio.

**Reply by mlichtblau** [reply]
Thank you for the clarification, that already helped to make a more informed decision.

> the overhead will be around 0.1%.

The overhead of an unblocked thread is 0.1%, but in the case I described the CPU time without the profiler would be 0 and increasing that to 100 microseconds is an infinite relative increase. If you have multiple thousands threads you expect to be sleeping for minutes, hours or days suddenly consuming 0.1% CPU time you end up with multiple cores being blocked only for recording the profile and restarting the `waitpid` syscall.

>  Is it a problem? I don't think so.

I guess this highly depends on the application, so I understand this is not necessarily a problem of the profiler0. In our case I guess we can live with CPU profiling without wall clock profiling. Thanks for the clarification!

**Reply by mlichtblau** [reply]
As in our case the `process reaper` threads in Question are spawned by the JDK I think it could be helpful for all users to have a workaround if one exists.

**Reply by apangin** [reply]
Do you say you have thousands threads sleeping in `waitpid` at the same time? 😲
Okay - in this case, filtering (#766) or increasing the sampling interval may help.
I have an idea how to improve wall clock profiling for threads that are mostly idle - I'll experiment with that.

**Reply by mlichtblau** [reply]
> Do you say you have thousands threads sleeping in waitpid at the same time?

Yes, spawning many child processes creates many sleeping threads.
Thanks for the recommendations, I'll be interested in what you come up with.
